## Initialisation

In [2]:
%load_ext sql

In [3]:
%sql mysql+mysqldb://root:root@some-mysql:3306/sakila

Connecting to 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

In [4]:
%config SqlMagic.style = '_DEPRECATED_DEFAULT'
%config SqlMagic.displaylimit = 300

## TP2 - première partie

Il s'agit simplement de revoir les bases de SQL à l'aide d'un ensemble de requêtes sur la base de données `sakila`.

Vous pouvez ajouter des cellules de code pour explorer la base si besoin, sinon vous pouvez utiliser l'extension Database Client installée plus tôt.

In [ ]:
%sql show tables
%config SqlMagic.displaylimit = 10

### Ex 1

Récupérer l'ID, le prénom et le nom de tous les acteurs. Trier le résultat suivant le nom puis le prénom (200 tuples).

In [6]:
%%sql
SELECT actor_id, first_name, last_name FROM actor ORDER BY last_name ASC, first_name

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

200 rows affected.

+----------+-------------+--------------+
| actor_id |  first_name |  last_name   |
+----------+-------------+--------------+
|    58    |  CHRISTIAN  |    AKROYD    |
|   182    |    DEBBIE   |    AKROYD    |
|    92    |   KIRSTEN   |    AKROYD    |
|   118    |     CUBA    |    ALLEN     |
|   145    |     KIM     |    ALLEN     |
|   194    |    MERYL    |    ALLEN     |
|    76    |   ANGELINA  |   ASTAIRE    |
|   112    |   RUSSELL   |    BACALL    |
|   190    |    AUDREY   |    BAILEY    |
|    67    |   JESSICA   |    BAILEY    |
|   115    |   HARRISON  |     BALE     |
|   187    |    RENEE    |     BALL     |
|    47    |    JULIA    |  BARRYMORE   |
|   158    |    VIVIEN   |   BASINGER   |
|   174    |   MICHAEL   |    BENING    |
|   124    |   SCARLETT  |    BENING    |
|    14    |    VIVIEN   |    BERGEN    |
|   121    |     LIZA    |   BERGMAN    |
|    91    | CHRISTOPHER |    BERRY     |
|    60    |    HENRY    |    BERRY     |
|    12    |     KARL    |    BERRY     |
|   189    |     CUBA    |    BIRCH     |
|    25    |    KEVIN    |    BLOOM     |
|   185    |   MICHAEL   |    BOLGER    |
|    37    |     VAL     |    BOLGER    |
|    98    |    CHRIS    |   BRIDGES    |
|    39    |    GOLDIE   |    BRODY     |
|   159    |    LAURA    |    BRODY     |
|   167    |   LAURENCE  |   BULLOCK    |
|    40    |    JOHNNY   |     CAGE     |
|    11    |     ZERO    |     CAGE     |
|   181    |   MATTHEW   |    CARREY    |
|    86    |     GREG    |   CHAPLIN    |
|    3     |      ED     |    CHASE     |
|   176    |     JON     |    CHASE     |
|   183    |   RUSSELL   |    CLOSE     |
|    16    |     FRED    |   COSTNER    |
|   129    |    DARYL    |   CRAWFORD   |
|    26    |     RIP     |   CRAWFORD   |
|    49    |     ANNE    |    CRONYN    |
|   104    |   PENELOPE  |    CRONYN    |
|   105    |    SIDNEY   |    CROWE     |
|    57    |     JUDE    |    CRUISE    |
|    80    |    RALPH    |     CRUZ     |
|    81    |   SCARLETT  |    DAMON     |
|    4     |   JENNIFER  |    DAVIS     |
|   110    |    SUSAN    |    DAVIS     |
|   101    |    SUSAN    |    DAVIS     |
|    48    |   FRANCES   |  DAY-LEWIS   |
|    35    |     JUDY    |     DEAN     |
|   143    |    RIVER    |     DEAN     |
|   148    |    EMILY    |     DEE      |
|   138    |   LUCILLE   |     DEE      |
|   107    |     GINA    |  DEGENERES   |
|    41    |    JODIE    |  DEGENERES   |
|   166    |     NICK    |  DEGENERES   |
|    89    |   CHARLIZE  |    DENCH     |
|   123    |   JULIANNE  |    DENCH     |
|   160    |    CHRIS    |     DEPP     |
|   100    |   SPENCER   |     DEPP     |
|   109    |  SYLVESTER  |     DERN     |
|   173    |     ALAN    |   DREYFUSS   |
|    36    |     BURT    |   DUKAKIS    |
|   188    |     ROCK    |   DUKAKIS    |
|   106    |   GROUCHO   |    DUNST     |
|    19    |     BOB     |   FAWCETT    |
|   199    |    JULIA    |   FAWCETT    |
|    10    |  CHRISTIAN  |    GABLE     |
|   165    |      AL     |   GARLAND    |
|   184    |   HUMPHREY  |   GARLAND    |
|   127    |    KEVIN    |   GARLAND    |
|   154    |    MERYL    |    GIBSON    |
|    46    |    PARKER   |   GOLDBERG   |
|   139    |     EWAN    |   GOODING    |
|   191    |   GREGORY   |   GOODING    |
|    71    |     ADAM    |    GRANT     |
|   179    |      ED     |   GUINESS    |
|    1     |   PENELOPE  |   GUINESS    |
|    90    |     SEAN    |   GUINESS    |
|    32    |     TIM     |   HACKMAN    |
|   175    |   WILLIAM   |   HACKMAN    |
|   152    |     BEN     |    HARRIS    |
|   141    |     CATE    |    HARRIS    |
|    56    |     DAN     |    HARRIS    |
|    97    |     MEG     |    HAWKE     |
|   151    |   GEOFFREY  |    HESTON    |
|   169    |   KENNETH   |   HOFFMAN    |
|    79    |     MAE     |   HOFFMAN    |
|    28    |    WOODY    |   HOFFMAN    |
|   161    |    HARVEY   |     HOPE     |
|   134    |     GENE    |   HOPKINS    |
|   113    |    MORGAN   |   HOPKINS    |
|    50   

### Ex 2

Récupérer l'ID, le prénom et le nom de tous les acteurs dont le nom est 'WILLIAMS' ou 'DAVIS' (6 tuples). Y a-t-il un doublon ?

In [7]:
%%sql
SELECT actor_id, first_name, last_name FROM actor WHERE last_name IN ('WILLIAMS', 'DAVIS');

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

6 rows affected.

+----------+------------+-----------+
| actor_id | first_name | last_name |
+----------+------------+-----------+
|    4     |  JENNIFER  |   DAVIS   |
|   101    |   SUSAN    |   DAVIS   |
|   110    |   SUSAN    |   DAVIS   |
|    72    |    SEAN    |  WILLIAMS |
|   137    |   MORGAN   |  WILLIAMS |
|   172    |  GROUCHO   |  WILLIAMS |
+----------+------------+-----------+

### Ex 3

Ecrire une requête portant sur la table `rental` qui retourne l'ID des clients qui ont loué un film le 25 mai 2005, de sorte qu'il n'y ait qu'une ligne par client.

Conseil :

*   utiliser l'attribut `rental_date` (quel est son type ?) et la fonction `date()` qui permet d'ignorer l'heure ;
*   le format d'une date est YYYY-MM-DD.

In [8]:
%%sql
SELECT DISTINCT customer_id FROM rental WHERE rental_date BETWEEN '2005-05-25 00:00:00' AND '2005-05-25 23:59:59';

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

129 rows affected.

+-------------+
| customer_id |
+-------------+
|     126     |
|     399     |
|     142     |
|     261     |
|     334     |
|     446     |
|     319     |
|     316     |
|     575     |
|      19     |
|     456     |
|     185     |
|     388     |
|     509     |
|     438     |
|     350     |
|      37     |
|     371     |
|     301     |
|     232     |
|      44     |
|     430     |
|     369     |
|     230     |
|     272     |
|     597     |
|     484     |
|      88     |
|     535     |
|     302     |
|     207     |
|     413     |
|     174     |
|     523     |
|     532     |
|     436     |
|      7      |
|      35     |
|     282     |
|     498     |
|      18     |
|     256     |
|     507     |
|     569     |
|     291     |
|     131     |
|     511     |
|      6      |
|     323     |
|     408     |
|     470     |
|     250     |
|     419     |
|     383     |
|     368     |
|     346     |
|      86     |
|     119     |
|     120     |
|     305     |
|      73     |
|     100     |
|      48     |
|     391     |
|     265     |
|     510     |
|      1      |
|     451     |
|     135     |
|     245     |
|     314     |
|     286     |
|     427     |
|     492     |
|     414     |
|     266     |
|     331     |
|      53     |
|     499     |
|      25     |
|     267     |
|     288     |
|     197     |
|     400     |
|      49     |
|     263     |
|     269     |
|     208     |
|     468     |
|     343     |
|     384     |
|     310     |
|     108     |
|     196     |
|     317     |
|     242     |
|     503     |
|     227     |
|     500     |
|      93     |
|     506     |
|     455     |
|     524     |
|      51     |
|     365     |
|     405     |
|     273     |
|      43     |
|     246     |
|     439     |
|      94     |
|      40     |
|      23     |
|      56     |
|     325     |
|     479     |
|     237     |
|     222     |
|     304     |
|     504     |
|     560     |
|     586     |
|     257     |
|     354     |
|      89     |
|     472     |
|     297     |
|     357     |
|      82     |
+-------------+

In [ ]:
%config SqlMagic.displaylimit = None

### Ex 4

Compléter les "blancs" (numérotés <1>, <2>…) dans la requête de façon à obtenir le résultat suivant :

| email                                 | return_date         |
|---------------------------------------|---------------------|
| DANIEL.CABRAL@sakilacustomer.org      | 2005-06-23 22:00:38 |
| TERRANCE.ROUSH@sakilacustomer.org     | 2005-06-23 21:53:46 |
| MIRIAM.MCKINNEY@sakilacustomer.org    | 2005-06-21 17:12:08 |
| GWENDOLYN.MAY@sakilacustomer.org      | 2005-06-20 02:40:27 |
| JEANETTE.GREENE@sakilacustomer.org    | 2005-06-19 23:26:46 |
| HERMAN.DEVORE@sakilacustomer.org      | 2005-06-19 03:20:09 |
| JEFFERY.PINSON@sakilacustomer.org     | 2005-06-18 21:37:33 |
| MATTHEW.MAHAN@sakilacustomer.org      | 2005-06-18 05:18:58 |
| MINNIE.ROMERO@sakilacustomer.org      | 2005-06-18 01:58:34 |
| SONIA.GREGORY@sakilacustomer.org      | 2005-06-17 21:44:11 |
| TERRENCE.GUNDERSON@sakilacustomer.org | 2005-06-17 05:28:35 |
| ELMER.NOE@sakilacustomer.org          | 2005-06-17 02:11:13 |
| JOYCE.EDWARDS@sakilacustomer.org      | 2005-06-16 21:00:26 |
| AMBER.DIXON@sakilacustomer.org        | 2005-06-16 04:02:56 |
| CHARLES.KOWALSKI@sakilacustomer.org   | 2005-06-16 02:26:34 |
| CATHERINE.CAMPBELL@sakilacustomer.org | 2005-06-15 20:43:03 |

Conseil :

*   voir clause `ORDER BY` ;
*   lors de la création d'un alias, le mot-clé `AS` est optionnel (`customer AS c` ou `customer c`).

In [9]:
%%sql # ex 4

SELECT c.email, r.return_date
FROM customer c
  INNER JOIN rental r
  ON c.customer_id = r.customer_id
WHERE date(r.rental_date) = '2005-06-14'
ORDER BY r.return_date DESC

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

16 rows affected.

+---------------------------------------+---------------------+
|                 email                 |     return_date     |
+---------------------------------------+---------------------+
|    DANIEL.CABRAL@sakilacustomer.org   | 2005-06-23 22:00:38 |
|   TERRANCE.ROUSH@sakilacustomer.org   | 2005-06-23 21:53:46 |
|   MIRIAM.MCKINNEY@sakilacustomer.org  | 2005-06-21 17:12:08 |
|    GWENDOLYN.MAY@sakilacustomer.org   | 2005-06-20 02:40:27 |
|   JEANETTE.GREENE@sakilacustomer.org  | 2005-06-19 23:26:46 |
|    HERMAN.DEVORE@sakilacustomer.org   | 2005-06-19 03:20:09 |
|   JEFFERY.PINSON@sakilacustomer.org   | 2005-06-18 21:37:33 |
|    MATTHEW.MAHAN@sakilacustomer.org   | 2005-06-18 05:18:58 |
|    MINNIE.ROMERO@sakilacustomer.org   | 2005-06-18 01:58:34 |
|    SONIA.GREGORY@sakilacustomer.org   | 2005-06-17 21:44:11 |
| TERRENCE.GUNDERSON@sakilacustomer.org | 2005-06-17 05:28:35 |
|      ELMER.NOE@sakilacustomer.org     | 2005-06-17 02:11:13 |
|    JOYCE.EDWARDS@sakilacustomer.org   | 2005-06-16 21:00:26 |
|     AMBER.DIXON@sakilacustomer.org    | 2005-06-16 04:02:56 |
|  CHARLES.KOWALSKI@sakilacustomer.org  | 2005-06-16 02:26:34 |
| CATHERINE.CAMPBELL@sakilacustomer.org | 2005-06-15 20:43:03 |
+---------------------------------------+---------------------+

### Ex 5

Compléter les "blancs" (numérotés <1>, <2>…) dans la requête de façon à obtenir le résultat suivant :

| first_name | last_name | address                | city           |
|------------|-----------|------------------------|----------------|
| PATRICIA   | JOHNSON   | 1121 Loja Avenue       | San Bernardino |
| BETTY      | WHITE     | 770 Bydgoszcz Avenue   | Citrus Heights |
| ALICE      | STEWART   | 1135 Izumisano Parkway | Fontana        |
| ROSA       | REYNOLDS  | 793 Cam Ranh Avenue    | Lancaster      |
| RENEE      | LANE      | 533 al-Ayn Boulevard   | Compton        |
| KRISTIN    | JOHNSTON  | 226 Brest Manor        | Sunnyvale      |
| CASSANDRA  | WALTERS   | 920 Kumbakonam Loop    | Salinas        |
| JACOB      | LANCE     | 1866 al-Qatif Avenue   | El Monte       |
| RENE       | MCALISTER | 1895 Zhezqazghan Drive | Garden Grove   |

In [10]:
%%sql # ex5

SELECT c.first_name, c.last_name, a.address, ct.city
FROM customer c
  INNER JOIN address a
  ON c.address_id = a.address_id
  INNER JOIN city ct
  ON a.city_id = ct.city_id
WHERE a.district = 'California';

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

9 rows affected.

+------------+-----------+------------------------+----------------+
| first_name | last_name |        address         |      city      |
+------------+-----------+------------------------+----------------+
|  PATRICIA  |  JOHNSON  |    1121 Loja Avenue    | San Bernardino |
|   BETTY    |   WHITE   |  770 Bydgoszcz Avenue  | Citrus Heights |
|   ALICE    |  STEWART  | 1135 Izumisano Parkway |    Fontana     |
|    ROSA    |  REYNOLDS |  793 Cam Ranh Avenue   |   Lancaster    |
|   RENEE    |    LANE   |  533 al-Ayn Boulevard  |    Compton     |
|  KRISTIN   |  JOHNSTON |    226 Brest Manor     |   Sunnyvale    |
| CASSANDRA  |  WALTERS  |  920 Kumbakonam Loop   |    Salinas     |
|   JACOB    |   LANCE   |  1866 al-Qatif Avenue  |    El Monte    |
|    RENE    | MCALISTER | 1895 Zhezqazghan Drive |  Garden Grove  |
+------------+-----------+------------------------+----------------+

Ecrire la même requête sans le join

In [12]:
%%sql # ex5

SELECT c.first_name, c.last_name, a.address, ct.city
FROM customer c, address a, city ct
WHERE c.address_id = a.address_id
    AND a.city_id = ct.city_id
    AND a.district = 'California';

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

/opt/conda/lib/python3.12/site-packages/sql/connection/connection.py:866: JupySQLRollbackPerformed: Found invalid transaction. JupySQL executed a ROLLBACK operation.
  warnings.warn(


9 rows affected.

+------------+-----------+------------------------+----------------+
| first_name | last_name |        address         |      city      |
+------------+-----------+------------------------+----------------+
|  PATRICIA  |  JOHNSON  |    1121 Loja Avenue    | San Bernardino |
|   BETTY    |   WHITE   |  770 Bydgoszcz Avenue  | Citrus Heights |
|   ALICE    |  STEWART  | 1135 Izumisano Parkway |    Fontana     |
|    ROSA    |  REYNOLDS |  793 Cam Ranh Avenue   |   Lancaster    |
|   RENEE    |    LANE   |  533 al-Ayn Boulevard  |    Compton     |
|  KRISTIN   |  JOHNSTON |    226 Brest Manor     |   Sunnyvale    |
| CASSANDRA  |  WALTERS  |  920 Kumbakonam Loop   |    Salinas     |
|   JACOB    |   LANCE   |  1866 al-Qatif Avenue  |    El Monte    |
|    RENE    | MCALISTER | 1895 Zhezqazghan Drive |  Garden Grove  |
+------------+-----------+------------------------+----------------+

### Ex 6

Ecrire une requête qui retourne le titre de tous les films dans lequel joue un acteur prénommé 'JOHN' (29 tuples).

In [13]:
%%sql
SELECT f.title
    FROM film f
    INNER JOIN film_actor fa
    ON f.film_id = fa.film_id
    INNER JOIN actor a
    ON fa.actor_id = a.actor_id
    WHERE a.first_name = 'JOHN'

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

29 rows affected.

+---------------------------+
|           title           |
+---------------------------+
|      ALLEY EVOLUTION      |
|       BEVERLY OUTLAW      |
|       CANDLES GRAPES      |
|      CLEOPATRA DEVIL      |
|     COLOR PHILADELPHIA    |
|       CONQUERER NUTS      |
|      DAUGHTER MADIGAN     |
|    GLEAMING JAWBREAKER    |
|      GOLDMINE TYCOON      |
|         HOME PITY         |
|     INTERVIEW LIAISONS    |
|      ISHTAR ROCKETEER     |
|        JAPANESE RUN       |
|        JERSEY SASSY       |
|         LUKE MUMMY        |
|        MILLION ACE        |
|     MONSTER SPARTACUS     |
|       NAME DETECTIVE      |
|     NECKLACE OUTBREAK     |
|       NEWSIES STORY       |
|        PET HAUNTING       |
|      PIANIST OUTFIELD     |
|      PINOCCHIO SIMON      |
|    PITTSBURGH HUNCHBACK   |
|        QUILLS BULL        |
|      RAGING AIRPLANE      |
|       ROXANNE REBEL       |
| SATISFACTION CONFIDENTIAL |
|        SONG HEDWIG        |
+---------------------------+

La même requête sans les JOIN ?

In [14]:
%%sql
SELECT f.title
    FROM film f, film_actor fa, actor a
    WHERE f.film_id = fa.film_id
        AND fa.actor_id = a.actor_id
        AND a.first_name = 'JOHN'

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

29 rows affected.

+---------------------------+
|           title           |
+---------------------------+
|      ALLEY EVOLUTION      |
|       BEVERLY OUTLAW      |
|       CANDLES GRAPES      |
|      CLEOPATRA DEVIL      |
|     COLOR PHILADELPHIA    |
|       CONQUERER NUTS      |
|      DAUGHTER MADIGAN     |
|    GLEAMING JAWBREAKER    |
|      GOLDMINE TYCOON      |
|         HOME PITY         |
|     INTERVIEW LIAISONS    |
|      ISHTAR ROCKETEER     |
|        JAPANESE RUN       |
|        JERSEY SASSY       |
|         LUKE MUMMY        |
|        MILLION ACE        |
|     MONSTER SPARTACUS     |
|       NAME DETECTIVE      |
|     NECKLACE OUTBREAK     |
|       NEWSIES STORY       |
|        PET HAUNTING       |
|      PIANIST OUTFIELD     |
|      PINOCCHIO SIMON      |
|    PITTSBURGH HUNCHBACK   |
|        QUILLS BULL        |
|      RAGING AIRPLANE      |
|       ROXANNE REBEL       |
| SATISFACTION CONFIDENTIAL |
|        SONG HEDWIG        |
+---------------------------+

### Ex 7

Ecrire une requête qui retourne les couples d'adresses distinctes (précisément les rues et numéros de rue) qui sont dans la même ville (4 tuples).

Conseil : utiliser une jointure de la table des adresses avec elle même pour retourner des doublets d'adresses dans la même ville.

In [20]:
%%sql
SELECT DISTINCT a.address, b.address
FROM address a
INNER JOIN address b
ON a.city_id = b.city_id
WHERE a.address != b.address AND a.address_id < b.address_id

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

4 rows affected.

+--------------------+----------------------+
|      address       |      address_1       |
+--------------------+----------------------+
| 47 MySakila Drive  |  23 Workhaven Lane   |
| 28 MySQL Boulevard | 1411 Lillydale Drive |
| 1497 Yuzhou Drive  |  548 Uruapan Street  |
| 587 Benguela Manor |   43 Vilnius Manor   |
+--------------------+----------------------+

### Ex 8

Ecrire une requête qui compte le nombre de tuples dans la table `payment` (16 044).

Conseil : utiliser la fonction `count`

In [21]:
%%sql
SELECT COUNT(*) FROM payment

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

1 rows affected.

+----------+
| COUNT(*) |
+----------+
|  16044   |
+----------+

Modifier la requête précédente afin de compter le nombre de paiements effectués par chaque client et y associer l'ID du client et le total de ses paiements.

| customer_id | nb_payment | total_payment |
|-------------|------------|---------------|
| 1           | 32         | 118.68        |
| 2           | 27         | 128.73        |
| 3           | 26         | 135.74        |
| 4           | 22         | 81.78         |
| 5           | 38         | 144.62        |
| ...         | ...        | ...           |
| 595         | 30         | 117.70        |
| 596         | 28         | 96.72         |
| 597         | 25         | 99.75         |
| 598         | 22         | 83.78         |
| 599         | 19         | 83.81         |

Conseil : utiliser la fonction `sum` et la clause `GROUP BY`.

In [22]:
%%sql
SELECT customer_id, COUNT(amount), SUM(amount) FROM payment GROUP BY customer_id

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

599 rows affected.

+-------------+---------------+-------------+
| customer_id | COUNT(amount) | SUM(amount) |
+-------------+---------------+-------------+
|      1      |       32      |    118.68   |
|      2      |       27      |    128.73   |
|      3      |       26      |    135.74   |
|      4      |       22      |    81.78    |
|      5      |       38      |    144.62   |
|      6      |       28      |    93.72    |
|      7      |       33      |    151.67   |
|      8      |       24      |    92.76    |
|      9      |       23      |    89.77    |
|      10     |       25      |    99.75    |
|      11     |       24      |    106.76   |
|      12     |       28      |    103.72   |
|      13     |       27      |    131.73   |
|      14     |       28      |    117.72   |
|      15     |       32      |    134.68   |
|      16     |       28      |    118.72   |
|      17     |       21      |    98.79    |
|      18     |       22      |    91.78    |
|      19     |       24      |    125.76   |
|      20     |       30      |    115.70   |
|      21     |       35      |    155.65   |
|      22     |       22      |    113.78   |
|      23     |       30      |    119.70   |
|      24     |       25      |    95.75    |
|      25     |       29      |    115.71   |
|      26     |       34      |    152.66   |
|      27     |       31      |    126.69   |
|      28     |       32      |    111.68   |
|      29     |       36      |    140.64   |
|      30     |       34      |    123.66   |
|      31     |       26      |    104.74   |
|      32     |       29      |    127.71   |
|      33     |       21      |    91.79    |
|      34     |       24      |    89.76    |
|      35     |       32      |    129.68   |
|      36     |       27      |    92.73    |
|      37     |       23      |    95.77    |
|      38     |       34      |    127.66   |
|      39     |       29      |    141.71   |
|      40     |       27      |    110.73   |
|      41     |       25      |    118.75   |
|      42     |       30      |    117.70   |
|      43     |       24      |    99.76    |
|      44     |       26      |    114.74   |
|      45     |       27      |    127.73   |
|      46     |       34      |    142.66   |
|      47     |       22      |    108.78   |
|      48     |       17      |    76.83    |
|      49     |       28      |    130.72   |
|      50     |       35      |    169.65   |
|      51     |       33      |    138.67   |
|      52     |       29      |    107.71   |
|      53     |       30      |    115.70   |
|      54     |       29      |    128.71   |
|      55     |       22      |    100.78   |
|      56     |       30      |    135.70   |
|      57     |       28      |    114.72   |
|      58     |       27      |    115.73   |
|      59     |       27      |    133.73   |
|      60     |       25      |    98.75    |
|      61     |       14      |    58.86    |
|      62     |       23      |    84.77    |
|      63     |       25      |    112.75   |
|      64     |       33      |    100.67   |
|      65     |       22      |    103.78   |
|      66     |       34      |    144.66   |
|      67     |       22      |    99.78    |
|      68     |       22      |    94.78    |
|      69     |       25      |    96.75    |
|      70     |       18      |    80.82    |
|      71     |       30      |    129.70   |
|      72     |       30      |    99.70    |
|      73     |       24      |    97.76    |
|      74     |       27      |    103.73   |
|      75     |       41      |    155.59   |
|      76     |       23      |    77.77    |
|      77     |       28      |    100.72   |
|      78     |       31      |    141.69   |
|      79     |       22      |    84.78    |
|      80     |       30      |    137.70   |
|      81     |       22      |    93.78    |
|      82     |       26      |    130.74   |
|      83     |       25      |    101.75   |
|      84     |       33      |    141.67   

Modifier la requête précédente afin de n'inclure que les clients qui ont fait au moins 40 paiements.

| customer_id | nb_payment | total_payment |
|-------------|------------|---------------|
| 75          | 41         | 155.59        |
| 144         | 42         | 195.58        |
| 148         | 46         | 216.54        |
| 197         | 40         | 154.60        |
| 236         | 42         | 175.58        |
| 469         | 40         | 177.60        |
| 526         | 45         | 221.55        |

Conseil : voir clause `HAVING`.

In [23]:
%%sql
SELECT customer_id, COUNT(amount), SUM(amount) FROM payment GROUP BY customer_id HAVING COUNT(amount) > 40

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

5 rows affected.

+-------------+---------------+-------------+
| customer_id | COUNT(amount) | SUM(amount) |
+-------------+---------------+-------------+
|      75     |       41      |    155.59   |
|     144     |       42      |    195.58   |
|     148     |       46      |    216.54   |
|     236     |       42      |    175.58   |
|     526     |       45      |    221.55   |
+-------------+---------------+-------------+

### Ex 9

Ecrire une requête qui retourne les titres des films d'action `category.name = 'Action'` (64 tuples de 'AMADEUS HOLY' à 'WORST BANGER').

Faites cette requête de 3 façons :

* avec des jointures 
* en filtrant la relation `film` à l'aide d'une sous-requête qui retourne l'ID de tous les films d'action
* sans jointure et sans filtre

In [24]:
%%sql
SELECT f.title
    FROM film f 
    JOIN film_category fc
    ON f.film_id = fc.film_id
    JOIN category c
    ON fc.category_id = c.category_id
    WHERE c.name = 'Action'

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

64 rows affected.

+-------------------------+
|          title          |
+-------------------------+
|       AMADEUS HOLY      |
|     AMERICAN CIRCUS     |
|    ANTITRUST TOMATOES   |
|      ARK RIDGEMONT      |
|   BAREFOOT MANCHURIAN   |
|       BERETS AGENT      |
|      BRIDE INTRIGUE     |
|      BULL SHAWSHANK     |
|     CADDYSHACK JEDI     |
|     CAMPUS REMEMBER     |
|    CASUALTIES ENCINO    |
|      CELEBRITY HORN     |
|     CLUELESS BUCKET     |
|       CROW GREASE       |
|       DANCES NONE       |
|       DARKO DORADO      |
|      DARN FORRESTER     |
|       DEVIL DESIRE      |
|       DRAGON SQUAD      |
|       DREAM PICKUP      |
|   DRIFTER COMMANDMENTS  |
|      EASY GLADIATOR     |
| ENTRAPMENT SATISFACTION |
|      EXCITEMENT EVE     |
|     FANTASY TROOPERS    |
|    FIREHOUSE VIETNAM    |
|     FOOL MOCKINGBIRD    |
|       FORREST SONS      |
|       GLASS DYING       |
|      GOSFORD DONNIE     |
|    GRAIL FRANKENSTEIN   |
|    HANDICAP BOONDOCK    |
|     HILLS NEIGHBORS     |
|      KISSING DOLLS      |
|      LAWRENCE LOVE      |
|       LORD ARIZONA      |
|        LUST LOCK        |
|    MAGNOLIA FORRESTER   |
|    MIDNIGHT WESTWARD    |
|       MINDS TRUMAN      |
|  MOCKINGBIRD HOLLYWOOD  |
|    MONTEZUMA COMMAND    |
|       PARK CITIZEN      |
|      PATRIOT ROMAN      |
|      PRIMARY GLASS      |
|     QUEST MUSSOLINI     |
|       REAR TRADING      |
|   RINGS HEARTBREAKERS   |
|   RUGRATS SHAKESPEARE   |
|      SHRUNK DIVINE      |
|         SIDE ARK        |
|       SKY MIRACLE       |
|        SOUTH WAIT       |
|      SPEAKEASY DATE     |
|  STAGECOACH ARMAGEDDON  |
|        STORY SIDE       |
|     SUSPECTS QUILLS     |
|       TRIP NEWTON       |
|       TRUMAN CRAZY      |
|     UPRISING UPTOWN     |
|  WATERFRONT DELIVERANCE |
|      WEREWOLF LOLA      |
|       WOMEN DORADO      |
|       WORST BANGER      |
+-------------------------+

In [25]:
%%sql
SELECT title
    FROM film WHERE film_id IN
    (SELECT fc.film_id FROM film_category fc
    JOIN category c
    ON fc.category_id = c.category_id
    WHERE c.name = 'Action')

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

64 rows affected.

+-------------------------+
|          title          |
+-------------------------+
|       AMADEUS HOLY      |
|     AMERICAN CIRCUS     |
|    ANTITRUST TOMATOES   |
|      ARK RIDGEMONT      |
|   BAREFOOT MANCHURIAN   |
|       BERETS AGENT      |
|      BRIDE INTRIGUE     |
|      BULL SHAWSHANK     |
|     CADDYSHACK JEDI     |
|     CAMPUS REMEMBER     |
|    CASUALTIES ENCINO    |
|      CELEBRITY HORN     |
|     CLUELESS BUCKET     |
|       CROW GREASE       |
|       DANCES NONE       |
|       DARKO DORADO      |
|      DARN FORRESTER     |
|       DEVIL DESIRE      |
|       DRAGON SQUAD      |
|       DREAM PICKUP      |
|   DRIFTER COMMANDMENTS  |
|      EASY GLADIATOR     |
| ENTRAPMENT SATISFACTION |
|      EXCITEMENT EVE     |
|     FANTASY TROOPERS    |
|    FIREHOUSE VIETNAM    |
|     FOOL MOCKINGBIRD    |
|       FORREST SONS      |
|       GLASS DYING       |
|      GOSFORD DONNIE     |
|    GRAIL FRANKENSTEIN   |
|    HANDICAP BOONDOCK    |
|     HILLS NEIGHBORS     |
|      KISSING DOLLS      |
|      LAWRENCE LOVE      |
|       LORD ARIZONA      |
|        LUST LOCK        |
|    MAGNOLIA FORRESTER   |
|    MIDNIGHT WESTWARD    |
|       MINDS TRUMAN      |
|  MOCKINGBIRD HOLLYWOOD  |
|    MONTEZUMA COMMAND    |
|       PARK CITIZEN      |
|      PATRIOT ROMAN      |
|      PRIMARY GLASS      |
|     QUEST MUSSOLINI     |
|       REAR TRADING      |
|   RINGS HEARTBREAKERS   |
|   RUGRATS SHAKESPEARE   |
|      SHRUNK DIVINE      |
|         SIDE ARK        |
|       SKY MIRACLE       |
|        SOUTH WAIT       |
|      SPEAKEASY DATE     |
|  STAGECOACH ARMAGEDDON  |
|        STORY SIDE       |
|     SUSPECTS QUILLS     |
|       TRIP NEWTON       |
|       TRUMAN CRAZY      |
|     UPRISING UPTOWN     |
|  WATERFRONT DELIVERANCE |
|      WEREWOLF LOLA      |
|       WOMEN DORADO      |
|       WORST BANGER      |
+-------------------------+

In [26]:
%%sql
SELECT f.title
    FROM film f, film_category fc, category c
    WHERE f.film_id = fc.film_id
    AND fc.category_id = c.category_id
    AND c.name = 'Action'

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

64 rows affected.

+-------------------------+
|          title          |
+-------------------------+
|       AMADEUS HOLY      |
|     AMERICAN CIRCUS     |
|    ANTITRUST TOMATOES   |
|      ARK RIDGEMONT      |
|   BAREFOOT MANCHURIAN   |
|       BERETS AGENT      |
|      BRIDE INTRIGUE     |
|      BULL SHAWSHANK     |
|     CADDYSHACK JEDI     |
|     CAMPUS REMEMBER     |
|    CASUALTIES ENCINO    |
|      CELEBRITY HORN     |
|     CLUELESS BUCKET     |
|       CROW GREASE       |
|       DANCES NONE       |
|       DARKO DORADO      |
|      DARN FORRESTER     |
|       DEVIL DESIRE      |
|       DRAGON SQUAD      |
|       DREAM PICKUP      |
|   DRIFTER COMMANDMENTS  |
|      EASY GLADIATOR     |
| ENTRAPMENT SATISFACTION |
|      EXCITEMENT EVE     |
|     FANTASY TROOPERS    |
|    FIREHOUSE VIETNAM    |
|     FOOL MOCKINGBIRD    |
|       FORREST SONS      |
|       GLASS DYING       |
|      GOSFORD DONNIE     |
|    GRAIL FRANKENSTEIN   |
|    HANDICAP BOONDOCK    |
|     HILLS NEIGHBORS     |
|      KISSING DOLLS      |
|      LAWRENCE LOVE      |
|       LORD ARIZONA      |
|        LUST LOCK        |
|    MAGNOLIA FORRESTER   |
|    MIDNIGHT WESTWARD    |
|       MINDS TRUMAN      |
|  MOCKINGBIRD HOLLYWOOD  |
|    MONTEZUMA COMMAND    |
|       PARK CITIZEN      |
|      PATRIOT ROMAN      |
|      PRIMARY GLASS      |
|     QUEST MUSSOLINI     |
|       REAR TRADING      |
|   RINGS HEARTBREAKERS   |
|   RUGRATS SHAKESPEARE   |
|      SHRUNK DIVINE      |
|         SIDE ARK        |
|       SKY MIRACLE       |
|        SOUTH WAIT       |
|      SPEAKEASY DATE     |
|  STAGECOACH ARMAGEDDON  |
|        STORY SIDE       |
|     SUSPECTS QUILLS     |
|       TRIP NEWTON       |
|       TRUMAN CRAZY      |
|     UPRISING UPTOWN     |
|  WATERFRONT DELIVERANCE |
|      WEREWOLF LOLA      |
|       WOMEN DORADO      |
|       WORST BANGER      |
+-------------------------+

## TP2 - partie optionnelle

### 1. Définition d'une fenêtre d'analyse

#### 1.1 Définition via une fonction d'agrégation

Soit la requête suivante sur la base de données sakila, qui donne les ventes mensuelles de mai à août 2005 (la fonction `quarter` retourne le trimestre associé à une date et la fonction `monthname` le mois).

In [28]:
%%sql 
SELECT quarter(payment_date) quarter,
  monthname(payment_date) month_nm,
  sum(amount) monthly_sales
FROM payment
WHERE year(payment_date) = 2005
GROUP BY quarter(payment_date), monthname(payment_date); 

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

4 rows affected.

+---------+----------+---------------+
| quarter | month_nm | monthly_sales |
+---------+----------+---------------+
|    2    |   May    |    4823.44    |
|    2    |   June   |    9629.89    |
|    3    |   July   |    28368.91   |
|    3    |  August  |    24070.14   |
+---------+----------+---------------+

Pour comparer ces chiffres avec les meilleures ventes mensuelles et les meilleures ventes du trimestre on a besoin de créer des **fenêtres** qui vont regrouper des lignes sur des critères temporels (mais le principe peut s'appliquer à d'autres types de valeur) et y associer de nouvelles colonnes, ce qu'on va pouvoir faire en associant des **clauses** aux fonctions d'agrégation.

Une fenêtre est créée à l'aide d'une clause `over` combinée avec des clauses optionnelles en paramètre. L'absence de clauses optionnelles indique que toutes les données sont prises en compte.

In [34]:
%%sql 
SELECT quarter(payment_date) quarter,
  monthname(payment_date) month_nm,
  sum(amount) monthly_sales,
  max(sum(amount))
    over () max_overall_sales,
  max(sum(amount))
    over (partition by quarter(payment_date)) max_qrtr_sales
FROM payment
WHERE year(payment_date) = 2005
GROUP BY quarter(payment_date), monthname(payment_date);

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

4 rows affected.

+---------+----------+---------------+-------------------+----------------+
| quarter | month_nm | monthly_sales | max_overall_sales | max_qrtr_sales |
+---------+----------+---------------+-------------------+----------------+
|    2    |   May    |    4823.44    |      28368.91     |    9629.89     |
|    2    |   June   |    9629.89    |      28368.91     |    9629.89     |
|    3    |   July   |    28368.91   |      28368.91     |    28368.91    |
|    3    |  August  |    24070.14   |      28368.91     |    28368.91    |
+---------+----------+---------------+-------------------+----------------+

Quelles peuvent être ces clauses optionnelles :

- `partition by` définit une fenêtre par valeur de l'expression passée en paramètre ;
- `order by` ordonne les tuples de la fenêtre suivant l'expression passée en paramètre (avec la possibilité d'un ordre ascendant ou descendant avec les mots-clefs `ASC` et `DESC`) ;
- `rows` définit la fenêtre en fonction du tuple courant, par exemple, `rows between 1 preceding and 1 following` sélectionne une fenêtre de 3 tuples centrée sur le tuple courant ; `rows  unbounded preceding` sélectionne tous les tuples précédents, tuple courant inclus ;
- `range` définit la fenêtre suivant un domaine de valeurs, par exemple, `range between interval 3 day preceding and interval 3 day  following` définit une fenêtre de 7 jours autour d'une date qui apparaîtrait dans une clause `order by`.

#### 1.2 Définition via des fonctions autres

##### 1.2.1 Classement

La fonction `rank` associe un classement au tuple courant en fonction de la fenêtre qui lui est associée.

In [35]:
%%sql 
SELECT quarter(payment_date) quarter,
  monthname(payment_date) month_nm,
  sum(amount) monthly_sales,
  rank() over (order by sum(amount) desc) sales_rank
FROM payment
WHERE year(payment_date) = 2005
GROUP BY quarter(payment_date), monthname(payment_date)
ORDER BY quarter, month_nm;

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

4 rows affected.

+---------+----------+---------------+------------+
| quarter | month_nm | monthly_sales | sales_rank |
+---------+----------+---------------+------------+
|    2    |   June   |    9629.89    |     3      |
|    2    |   May    |    4823.44    |     4      |
|    3    |  August  |    24070.14   |     2      |
|    3    |   July   |    28368.91   |     1      |
+---------+----------+---------------+------------+

La quatrième colonne `sales_rank` donne ainsi le classement des ventes mensuelles. Il est par exemple possible de raffiner le classement par rapport à une fenêtre de temps, par exemple trimestrielle, en ajoutant une clause `partition by`.

In [36]:
%%sql 
SELECT quarter(payment_date) quart,
  monthname(payment_date) month_nm,
  sum(amount) monthly_sales,
    rank() over (
      partition by quarter(payment_date)
      order by sum(amount) desc
    ) qtr_sales_rank
FROM payment
WHERE year(payment_date) = 2005
GROUP BY quarter(payment_date), monthname(payment_date)
ORDER BY quart, month_nm;

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

4 rows affected.

+-------+----------+---------------+----------------+
| quart | month_nm | monthly_sales | qtr_sales_rank |
+-------+----------+---------------+----------------+
|   2   |   June   |    9629.89    |       1        |
|   2   |   May    |    4823.44    |       2        |
|   3   |  August  |    24070.14   |       2        |
|   3   |   July   |    28368.91   |       1        |
+-------+----------+---------------+----------------+

**Gestion des égalités**

Il y a différentes manières de traiter les égalités, comme le montre la requête suivante (la fonction `dense_rank` évite les trous dans le classement) :

In [37]:
%%sql 
SELECT customer_id, count(*) num_rentals,
  row_number() over (order by count(*) desc) row_number_rnk,
  rank() over (order by count(*) desc) rank_rnk,
  dense_rank() over (order by count(*) desc) dense_rank_rnk
FROM rental
GROUP BY customer_id
ORDER BY num_rentals DESC;

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

599 rows affected.

+-------------+-------------+----------------+----------+----------------+
| customer_id | num_rentals | row_number_rnk | rank_rnk | dense_rank_rnk |
+-------------+-------------+----------------+----------+----------------+
|     148     |      46     |       1        |    1     |       1        |
|     526     |      45     |       2        |    2     |       2        |
|     144     |      42     |       3        |    3     |       3        |
|     236     |      42     |       4        |    3     |       3        |
|      75     |      41     |       5        |    5     |       4        |
|     197     |      40     |       6        |    6     |       5        |
|     469     |      40     |       7        |    6     |       5        |
|     137     |      39     |       8        |    8     |       6        |
|     178     |      39     |       9        |    8     |       6        |
|     468     |      39     |       10       |    8     |       6        |
|      5      |      38     |       11       |    11    |       7        |
|     295     |      38     |       12       |    11    |       7        |
|     410     |      38     |       13       |    11    |       7        |
|     459     |      38     |       14       |    11    |       7        |
|     176     |      37     |       15       |    15    |       8        |
|     198     |      37     |       16       |    15    |       8        |
|     257     |      37     |       17       |    15    |       8        |
|     366     |      37     |       18       |    15    |       8        |
|      29     |      36     |       19       |    19    |       9        |
|     267     |      36     |       20       |    19    |       9        |
|     348     |      36     |       21       |    19    |       9        |
|     354     |      36     |       22       |    19    |       9        |
|     380     |      36     |       23       |    19    |       9        |
|     439     |      36     |       24       |    19    |       9        |
|      21     |      35     |       25       |    25    |       10       |
|      50     |      35     |       26       |    25    |       10       |
|      91     |      35     |       27       |    25    |       10       |
|     196     |      35     |       28       |    25    |       10       |
|     204     |      35     |       29       |    25    |       10       |
|     273     |      35     |       30       |    25    |       10       |
|     274     |      35     |       31       |    25    |       10       |
|     368     |      35     |       32       |    25    |       10       |
|     371     |      35     |       33       |    25    |       10       |
|     373     |      35     |       34       |    25    |       10       |
|     381     |      35     |       35       |    25    |       10       |
|     403     |      35     |       36       |    25    |       10       |
|     506     |      35     |       37       |    25    |       10       |
|      26     |      34     |       38       |    38    |       11       |
|      30     |      34     |       39       |    38    |       11       |
|      38     |      34     |       40       |    38    |       11       |
|      46     |      34     |       41       |    38    |       11       |
|      66     |      34     |       42       |    38    |       11       |
|     119     |      34     |       43       |    38    |       11       |
|     129     |      34     |       44       |    38    |       11       |
|     138     |      34     |       45       |    38    |       11       |
|     147     |      34     |       46       |    38    |       11       |
|     168     |      34     |       47       |    38    |       11       |
|     181     |      34     |       48       |    38    |       11       |
|     207     |      34     |       49       |    38    |       11       |
|     211     |      34     |       50       |    38    |       11       |
|     239     |      34  

##### 1.2.2 Comparaison d'une valeur avec les valeurs suivantes ou précédentes

Ce type de comparaison s'effectue avec les fonctions `lag` (qui définie une fenêtre sur les valeurs précédentes) et `lead` (qui définie une fenêtre sur les valeurs suivantes). Le premier paramètre donne la valeur à comparer et le deuxième (par défaut 1) la taille de la fenêtre.

Par exemple

In [38]:
%%sql
 
SELECT yearweek(payment_date) payment_week,
  sum(amount) week_total,
  lag(sum(amount), 1)
     over (order by yearweek(payment_date)) prev_wk_tot,
  lead(sum(amount), 1)
     over (order by yearweek(payment_date)) next_wk_tot
FROM payment
GROUP BY yearweek(payment_date)
ORDER BY payment_week;

Running query in 'mysql+mysqldb://root:***@some-mysql:3306/sakila'

11 rows affected.

+--------------+------------+-------------+-------------+
| payment_week | week_total | prev_wk_tot | next_wk_tot |
+--------------+------------+-------------+-------------+
|    200521    |  2846.19   |     None    |   1977.25   |
|    200522    |  1977.25   |   2846.19   |   5603.43   |
|    200524    |  5603.43   |   1977.25   |   4026.46   |
|    200525    |  4026.46   |   5603.43   |   8490.83   |
|    200527    |  8490.83   |   4026.46   |   5982.64   |
|    200528    |  5982.64   |   8490.83   |   11027.23  |
|    200530    |  11027.23  |   5982.64   |   8412.07   |
|    200531    |  8412.07   |   11027.23  |   10619.11  |
|    200533    |  10619.11  |   8412.07   |   7907.17   |
|    200534    |  7907.17   |   10619.11  |    514.18   |
|    200607    |   514.18   |   7907.17   |     None    |
+--------------+------------+-------------+-------------+

Closing mysql+mysqldb://root:***@some-mysql:3306/sakila

#### 1.3 Pour en savoir plus

Voir [https://dev.mysql.com/doc/refman/8.0/en/window-functions.html](https://dev.mysql.com/doc/refman/8.0/en/window-functions.html).

### Ex 1

Créer la table `sales_fact` avec

|   year_no  | month_no |  tot_sales  |
|------|----|-------|
| 2019 | 1  | 19228 |
| 2019 | 2  | 18554 |
| 2019 | 3  | 17325 |
| 2019 | 4  | 13221 |
| 2019 | 5  | 9964  |
| 2019 | 6  | 12658 |
| 2019 | 7  | 14233 |
| 2019 | 8  | 17342 |
| 2019 | 9  | 16853 |
| 2019 | 10 | 17121 |
| 2019 | 11 | 19095 |
| 2019 | 12 | 21436 |
| 2020 | 1  | 20347 |
| 2020 | 2  | 17434 |
| 2020 | 3  | 16225 |
| 2020 | 4  | 13853 |
| 2020 | 5  | 14589 |
| 2020 | 6  | 13248 |
| 2020 | 7  | 8728  |
| 2020 | 8  | 9378  |
| 2020 | 9  | 11467 |
| 2020 | 10 | 13842 |
| 2020 | 11 | 15742 |
| 2020 | 12 | 18636 |

In [ ]:
%%sql # creation table
CREATE TABLE sales_fact (
    ...
    PRIMARY KEY(...)
);

In [ ]:
%%sql
INSERT INTO ... VALUES ...;

Vérifier que le total des ventes est 370 519.

### Ex 2

Ecrire une requête qui ajoute une colonne qui classe les ventes sur toute la période ( 1 pour le meilleur classement et 24 pour le moins bon).

Modifier la requête pour que la dernière colonne calcule des classements par année.

### Ex 3

Ecrire une requête qui sélectionne les résultats de 2020 et ajoute une colonne qui pour chaque mois contient les résultats du mois précédent.

### Ex 5

JupySQL permet une intégration avec la célèbre bibliothèque Pandas. Pandas est une bibliothèque Python puissante et flexible utilisée pour la manipulation et l'analyse de données, offrant des structures comme les DataFrames pour gérer facilement des ensembles de données tabulaires.
Seaborn est une bibliothèque de visualisation de données basée sur Matplotlib, permettant de créer des graphiques statistiques facilement et avec des thèmes esthétiques intégrés pour l'analyse de données. Nous allons utiliser ces bibliothèque pour afficher des résultats.

Consultez la documentation de JupySQL et Seaborn et affichez une courbe présentant les résultats par jour, en groupant par catégorie de film.

Conseil : 

* utilisez la fonction seaborn `relplot`, vous pouvez utiliser les paramètres `hue`, `row` ou `col` pour regarder les détails par catégorie.

In [95]:
%%sql result << # ex 4
SELECT ...


In [96]:
import pandas as pd 
df = pd.DataFrame(result)

In [97]:
import seaborn as sns
sns.relplot(...)

### Ex 6

On voit des chutes de ventes sur certains jours. Lesquels sont-ils ? Préparer une requête employant la fonction `rank` pour les découvrir.

In [ ]:
%%sql 
SELECT ...;

In [ ]:
import seaborn as sns
sns.relplot(...)